st_sampler.py

```diff
@@ -0,0 +1,168 @@
+import random
+
+# given a walk as a list of vertices
+# returns a newwalk which has no loops
+# uses -1 as a invalid value
+def eraseLoops(walk):
+    result = [-1] * (len(walk) + 1)
+    ptr = 0
+    seen = {}
+    for v in walk:
+        if v in seen:
+            newptr = seen[v] # erase everything until this point
+            for i in range(newptr+1, ptr): # remove these node from seen
+                del seen[result[i]]
+            ptr = newptr + 1
+
+        else:
+            result[ptr] = v # add to result
+            seen[v] = ptr # record this location
+            ptr += 1
+        result[ptr] = -1 # do this to mark the next location
+
+    # copy the result into an array of correct size
```

Comment:
Hmm, why would the array not be of correct size? It doesn't seem like we are doing any filtering here either so I don't think the size of this array changes?